cu29/src/curuntime.rs

```diff
@@ -129,122 +148,221 @@ pub struct CuExecutionLoop {
                     .graph
-                    .edge_weight(EdgeIndex::new(config.get_dst_edges(id)[0]))
-                    .unwrap()
-                    .msg
-                    .clone(),
-            );
-            // get the node from where the message is coming from
-            let parent = config
-                .graph
-                .neighbors_directed(id.into(), Incoming)
-                .next()
-                .unwrap();
-            // Find the source of the incoming message
-            culist_input_index = find_output_index_from_nodeid(parent.index() as NodeId, &result);
-            CuTaskType::Sink
+                    .neighbors_directed(id.into(), Incoming)
+                    .collect();
+
+                for parent in parents {
+                    let index_type =
+                        find_output_index_type_from_nodeid(parent.index() as NodeId, &plan);
+                    eprintln!(
+                        "Found parent node: {:?} with index {:?}",
+                        parent, index_type
+                    );
+                    if let Some(index_type) = index_type {
+                        input_msg_indices_types.push(index_type);
+                    } else {
+                        eprintln!(
+                            "Parent node not in this branch, it will be picked up as a later pass."
+                        );
+                        // here do not add this node yet, wait for the other inputs to do it with all the inputs earliers in the copper list.
+                        return next_culist_output_index;
+                    }
+                }
+                output_msg_index_type = Some((
+                    next_culist_output_index,
+                    config
+                        .graph
+                        .edge_weight(EdgeIndex::new(config.get_src_edges(id)[0]))
+                        .unwrap()
+                        .msg
+                        .clone(),
+                ));
+                next_culist_output_index += 1;
+            }
+        }
+
+        // Sort the input messages by index
+        // It means that the tuple presented as input to the merging task
+        // depends on the order of *declaration* in the node section of the config file.
+        input_msg_indices_types.sort_by(|a, b| a.0.cmp(&b.0));
+
+        // Try to see if we did not already add this node to the plan
+        if let Some(pos) = plan.iter().position(|step| {
+            if let CuExecutionUnit::Step(ref s) = step {
+                s.node_id == id
+            } else {
+                false
+            }
+        }) {
+            // modify the existing step and put it back in the plan as the current step as it needs this subsequent output.
+            let mut step = plan.remove(pos);
+            if let CuExecutionUnit::Step(ref mut s) = step {
+                s.input_msg_indices_types = input_msg_indices_types;
+            }
+            plan.push(step);
```

Comment:
Why not modify the step in-place instead of popping and pushing it?

Reply:
oh because you need to put it at the end of the copper list to be sure all the tasks sending it messages has been called prior to it.